pkg/Library/CoreXenon/Framework/Binder.js

```diff
@@ -3,158 +3,105 @@
 };
 
 const addAtomSpecBindings = (bindings, atomId, atomSpec) => {
-  // unpack the objectId
-  const objectId = Id.sliceId(atomId, 0, -1);
-  // add bindings from the atom spec
-  map(atomSpec.bindings, (prop, bound) => addBoundProperty(bindings, objectId, atomId, prop, bound));  
-  // inputs
-  atomSpec.inputs?.forEach(prop => addInputBinding(bindings, quid(atomId, prop), atomId, prop));
-  // outputs
-  atomSpec.outputs?.forEach(prop => addOutputBinding(bindings, atomId, prop, quid(atomId, prop)));
-};
-
-const addBoundProperty = (bindings, objectId, atomId, inProp, bound) => {
-  // Array-ify simple-string 'bound' values
-  bound = typeof bound === 'string' ? [bound] : bound;
-  // for each value bound on this atom spec
-  bound?.forEach(binding => {
-    // create mapping from qualified state prop to input prop
-    addInputBinding(bindings, quid(objectId, binding), atomId, inProp);
-    // create mapping from output prop to qualified state prop
-    const [outId, outProp] = Id.splitId(binding);
-    addOutputBinding(bindings, quid(objectId, outId), outProp, quid(objectId, binding));
+  // each atom input receives from a state key
+  atomSpec.inputs?.forEach(prop => {
+    const key = Id.qualifyId(atomId, prop);
+    addBinding(bindings.input, key, key);
   });
-};
-
-export const addConnections = (layerId, connections, inputBindings) => {
-  // update input/outputBindings with cross-node connections
-  entries(connections).forEach(([propId, bound]) => {
-    const {id, prop} = Id.parsePropId(propId);
-    if (typeof bound === 'string') {
-      bound = [bound];
+  // unpack the node id
+  const nodeId = Id.sliceId(atomId, 0, -1);
+  // add input bindings from the atom spec
+  map(atomSpec.bindings, (prop, bounded) => {
+    if (!Array.isArray(bounded)) {
+      bounded = [bounded];
     }
-    // for each output target
-    bound?.forEach(bound => {
-      addInputBinding({inputBindings}, quid(layerId, bound), quid(layerId, id), prop);
+    const value = Id.qualifyId(atomId, prop);
+    addBinding(bindings.output, value, value);
+    bounded.forEach(bound => {
+      const key = Id.qualifyId(nodeId, bound);
+      addBinding(bindings.input, key, key);
+      addBinding(bindings.input, key, value);
     });
   });
-}
+  // each atom output sends to a state key
+  atomSpec.outputs?.forEach(prop => {
+    const key = Id.qualifyId(atomId, prop);
+    addBinding(bindings.output, key, key);
+  });
+};
 
-export const processOutput = (name, output, outputBindings) => {
-  // output object contains App-scoped key-value pairs.
-  // `outputBindings` shows how to remap property names 
-  // from Atom-scope to App-scope.
-  const map = outputBindings[name];
-  if (map) {
-    const scoped = scopeOutput(output, map);
-    if (keys(scoped).length) {
-      log('output', keys(output), 'bound to', keys(scoped), 'for', name);
-    }
-    return scoped;
+const addBinding = (bindings, key, value) => {
+  const bound = bindings[key] ??= [];
```

Comment:
i really like how input and output bindings now have the same structure 🙏

Reply:
Me too, but I wonder about the cost here:

https://github.com/NeonFlan/xenonjs/pull/93/files#diff-068878bcbd2157fb1c3438099b44b2f0258b28620c8878794c346a72c9a51008R82